libcontainerd/client_linux.go

```diff
@@ -456,30 +451,36 @@ func (clnt *client) Restore(containerID string, attachStdio StdioCallback, optio
 	// Get its last event
 	ev, eerr := clnt.getContainerLastEvent(containerID)
 	if err != nil || cont.Status == "Stopped" {
-		if err != nil && !strings.Contains(err.Error(), "container not found") {
-			// Legitimate error
-			return err
+		if err != nil {
+			logrus.Warnf("libcontainerd: failed to retrieve container %s state: %v", containerID, err)
 		}
-
-		if ev == nil {
-			if _, err := clnt.getContainer(containerID); err == nil {
-				// If ev is nil and the container is running in containerd,
-				// we already consumed all the event of the
-				// container, included the "exit" one.
-				// Thus we return to avoid overriding the Exit Code.
-				logrus.Warnf("libcontainerd: restore was called on a fully synced container (%s)", containerID)
-				return nil
-			}
-			// the container is not running so we need to fix the state within docker
-			ev = &containerd.Event{
-				Type:   StateExit,
-				Status: 1,
+		if ev != nil && ev.Pid != InitFriendlyName || ev.Type != StateExit {
+			// Wait a while for the exit event
+			timeout := time.After(10 * time.Second)
+			stop := false
```

Comment:
not a fan of such pattern, create helper and return from select if you don't like `break LABEL`